### DISCUSSION TITLE
GPIO Option A5 (Xiaomi Desk Lamp 1S vs. Xiaomi Desk Lamp Pro)

### DISCUSSION THREAD

**Discussion by zakaszewski** [arendst/Tasmota#14765] [Q&A] [answered]
Hi All,

can anyone provide me with info about what exactly "Option A5" does? It is automatically set on GPIO00 via template when configuring Tasmota for both my Xiaomi Mi Desk Lamp **1S**s (**MJTD01SYL**), which uses the same ESP32 that Xiaomi Mi Desk Lamp **Pro** (**MJTD02YL**) does, albeit with a different breakout. The latter also has a [Tasmota Template](https://templates.blakadder.com/xiaomi_MJTD02YL.html), which seems to be the ONLY reference to "Option A5" I could find.

Underlying issue: After successfully flashing tasmota32solo1 via usb2serial I found that there's a minor difference between Tasmota and the original firmware, namely in how the "knob/button/dial thingy" translates to light control when using aforementioned template:

| Action                                             | Original Firmware       | Tasmota                 |
|----------------------------------------------------|-------------------------|-------------------------|
| Button press                                       | Toggle light ON/OFF     | Toggle light ON/OFF     |
| Turn knob clockwise                                | Raise brightness        | **Lower** brightness        |
| Turn knob counterclockwise                         | Lower brightness        | **Raise** brightness        |
| Turn knob clockwise while button is pressed        | Raise color temperature | Raise color temperature |
| Turn knob counterclockwise while button is pressed | Lower color temperature | Lower color temperature |

A question to owners of a "tasmotized" Pro lamp: Do You see the same behaviour?

I already figured out that switching the assignment of GPIOs 26/27 from "Rotary B1/Rotary A1" to "Rotary A1/Rotary B1" (fully) inverts the turn behaviour, but that only moves the mismatch issue from "brightness" to "color temperature"; any other trial-and-error  changes to the GPIO config lead to either some features NOT working (e.g. knob controls being partially or completely disabled), or to erratic dimming/color temp changes. Also, probing the device as described in the [configuration procedure for new devices](https://tasmota.github.io/docs/Configuration-Procedure-for-New-Devices/) didn't bring any insight for me.

I imagine that this particular detail may be up to one's personal preference, so if there's a solution that involves compiling a custom binary, I'd be willing to try that; either way, I'll gladly add details on the procedure I followed as a new template. Meanwhile, I'll try to figure out how to sync both desk lamps with one another AND the corresponding ceiling light...

**Comment by barbudor** [comment, marked as answer]
"Option Ax" is a kind of setoption feature that can be added inside a template.
It doesn't actually make use of the GPIO, it just trigger a feature and could be assigned to any unused GPIO.

Unfortunately I don't own a xiaomi lamp and can't help on that topic

**Reply by sfromis** [reply]
I've got the ESP32 based "pro" [MJTD02YL](https://templates.blakadder.com/xiaomi_MJTD02YL.html) with the template from the repository and turning the knob clockwise goes from lower to higher brightness, or with it pushed at the same time, from higher (colder, blue-ish) to lower (warmer, yellow-ish) color temperature. As this appears to be the opposite direction for same effects, you should indeed be able to get to what I see by swapping the rotary gpios.

I did not use the lamp much before flashing Tasmota, and haven't compared behavior, or cared much about details. For me, adjusting color temperature is based on the visuals, seeing what happens, and turn the other way if relevant. The one getting the ESP32 based ones to work with Tasmota did notice that the color temperature direction is the opposite from original firmware, but did not think of this as an issue.

The source code describes "Option_A5" as "(Rotary) Enable Mi Desk emulation". This was a replacement for having a build time flag depending on rotary encoder type. It is not for configuring the behavior you see, but for successfully sensing the rotary encoder action.

The Tasmota driver code seems to use the outcome of the rotary encoder to control the internal color handling model, resulting in the directions you see. This was not done differently for the ESP32 based models, but possibly to match the similar but older ESP8266-based desk lamps (which I have no experience with). 

If you want to have fun with the code, [here](https://github.com/arendst/Tasmota/blob/development/tasmota/support_rotary.ino#L225-L255) is where the rotary inputs are used to control light actions.

**Reply by zakaszewski** [reply]
Thanks for the info and the detailed clarification! I'll look into the code and fiddle with it, hopfeully neither bricking the lamp nor burning down the house ;-)

**Reply by Sergey-SRG** [reply]
Hi! Maybe someone has the original firmware for Xiaomi Desk Lamp 1S?

**Reply by sfromis** [reply]
In general, do not expect such a "backup" to work on another device, as it is common for factory firmware to be locked per chip.
